### PR TITLE
Rebase EIP-4844 on Capella

### PR DESCRIPTION
**Pull request by Inphi**
Also adds a "testing" specification that allows withdrawals to be disabled during testing.

### CODE REVIEW

File: specs/eip4844/beacon-chain.md
Comment (hwwhww) [inline specs/eip4844/beacon-chain.md:276]:
```suggestion
        withdrawals_root=hash_tree_root(payload.withdrawals) if ENABLE_WITHDRAWALS else Bytes32(),  # [New in EIP-4844]
    )
```

File: specs/eip4844/beacon-chain.md
Comment (hwwhww) [inline specs/eip4844/beacon-chain.md:298]: That means we have to replace `@with_capella_and_later` decorator in tests with `@with_phases([CAPELLA])` if we treat `ENABLE_WITHDRAWALS == 0` as a constant.

And if we want to have the flexibility to test `ENABLE_WITHDRAWALS == 1`, we need to write a new decorator to check preset before running the test.

File: specs/eip4844/fork.md
Comment (hwwhww) [inline specs/eip4844/fork.md:70]: It's lowercase `capella` for the pyspec library.

**Review by hwwhww — COMMENTED** [review]
Had a quick scan, @Inphi well done on the rebasing!  

- Note that #3042 changed the consensus containers & epoch processing functions. So if we include it, you'd need to update that part.
- From pyspec's point of view, having `ENABLE_WITHDRAWALS` is kinda annoying 😬  but I'd defer to client devs if they really want this flag control in spec.

**Comment by terencechain** [comment]
> Had a quick scan, @Inphi well done on the rebasing!
> 
>     * Note that #3042 changed the consensus containers & epoch processing functions. So if we include it, you'd need to update that part.
> 
>     * From pyspec's point of view, having `ENABLE_WITHDRAWALS` is kinda annoying 😬  but I'd defer to client devs if they really want this flag control in spec.

Ditto. I don't think we should enshrine test-only flag in the spec. The decision on whether withdrawal is enabled or in the protocol for testing and testnet can be reached out of the band

File: specs/eip4844/beacon-chain.md
Comment (realbigsean) [reply specs/eip4844/beacon-chain.md:107]: Otherwise we're gonna have to create new objects that will never exist (an eip4844 block minus withdrawals)

**Comment by Inphi** [comment]
The goal is to align the behavior of disabling withdrawals across all clients. ENABLE_WITHDRAWALS will be removed eventually upon further testing. But right now, we're trying to avoid any withdrawal bugs or changes in the spec from influencing 4844.

File: specs/eip4844/beacon-chain.md
Comment (Inphi) [reply specs/eip4844/beacon-chain.md:107]: yeah we wanna retain the fields to make it easy for client devs to branch off their existing withdrawals implementation. I expect little to no bugs/changes to that implementation and specification.

Only thing we wanna avoid is triggering state transition logic for withdrawals since that's where they could most likely be different behaviors in client implementations.

**Comment by realbigsean** [comment]
In testnets we can avoid worrying too much about the impact of withdrawals on consensus by stubbing out these functions:

Epoch processing:
-`process_full_withdrawals`
-`process_partial_withdrawals`

Block processing:
- `process_withdrawals`
- `for_ops(body.bls_to_execution_changes)`

But I think we can figure this out on a per-testnet basis, by saying "target this spec commit but stub out these functions" rather than ingraining this in the spec here

I think it's simpler if we don't touch anything else, for example:
- Actually get the root of an empty withdrawals list
- Keep an empty withdrawals list in payload attributes
- Use EIP4844 + withdrawal types in all APIs

**Comment by Inphi** [comment]
@realbigsean Thanks for the suggestions. I've made the changes to the spec. `ENABLE_WITHDRAWALS` has been removed and updated the Testing section to specify how clients can disable withdrawals.

**Review by hwwhww — APPROVED** [review]
I managed to make withdrawal functions no-op & add basic tests, and find okay Wi-Fi somewhere 2km far from Machu Picchu. 😄

The functionalities/correctness LGTM now!

Edited: I'll defer to client devs to make the decision in testing. Although I personally found that the no-op stub is still somewhat annoying. 😅

**Review by djrtwo — CHANGES_REQUESTED** [review]
#3068 is merged. This is ready for a re-rebase @Inphi

**Comment by Inphi** [comment]
I've updated the PR for #3068. Reviewers please take another look.

File: specs/eip4844/p2p-interface.md
Comment (terencechain) [inline specs/eip4844/p2p-interface.md:144]: Potential merge conflict with #3089. I can fix that after this is merged

File: specs/eip4844/validator.md
Comment (terencechain) [inline specs/eip4844/validator.md:48]: `get_blobs_and_kzg_commitments` is still mentioned below, do we want to remove this?

**Review by terencechain — COMMENTED** [review]
Great work @Inphi ! Just one question

File: specs/eip4844/validator.md
Comment (Inphi) [reply specs/eip4844/validator.md]: Good catch. This was a bad merge from dev. I've added it back in. Thanks.

**Review by hwwhww — APPROVED** [review]
lgtm! :shipit:
 
I have a test vector generator branch (https://github.com/ethereum/consensus-specs/compare/pr3052-testgen) based on this PR. I'll open a PR once this one is merged.

File: specs/eip4844/beacon-chain.md
Comment (ralexstokes) [inline specs/eip4844/beacon-chain.md:139]: also, `excess_blobs` has become `excess_data_gas` with `uint256` type

this should be addressed here and in the `ExecutionPayload`

Comment (Inphi) [reply specs/eip4844/beacon-chain.md:139]: Thanks. Fixed.

File: specs/eip4844/beacon-chain.md
Comment (djrtwo) [inline specs/eip4844/beacon-chain.md:347]: This no-op works fine for clients?
I'm out of the loop, but I thought they would use the full functionality here (yes, complicating testing)

File: tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_bls_to_execution_change.py
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_bls_to_execution_change.py:48]: This should be capella and later, no?

Ah, the no-op change will break this test... If we do this, we need to remember to go back and change these all :/
When we were rebasing here, I didn't realize we were rebasing and mutating. How much will we be blocked on 4844 if we don't do the capella no-op within 4844 spec?

Comment (hwwhww) [reply tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_bls_to_execution_change.py:48]: My understanding is to prevent random semi-public testnet users/validators from invoking any withdrawals?

Agreed with this complicating testing but given that we have 3-4 in-discussion changes on Capella, it makes sense that EIP-4844 devs want to disable Capella now... 😭

Btw I also added the tests to test the Capella no-op. So we must undo it altogether.

File: specs/eip4844/beacon-chain.md
Comment (Inphi) [reply specs/eip4844/beacon-chain.md:347]: We want to avoid submitted withdrawals from interfering with the testnet. By no-op'ing these functions, withdrawals won't have any effect.

File: tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_bls_to_execution_change.py
Comment (Inphi) [reply tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_bls_to_execution_change.py:48]: > My understanding is to prevent random semi-public testnet users/validators from invoking any withdrawals?

Yes it's exactly this.

File: specs/eip4844/p2p-interface.md
Comment (hwwhww) [reply specs/eip4844/p2p-interface.md:146]: fixed in https://github.com/ethereum/consensus-specs/pull/3052/commits/f1d4c9047a70c3a321cb0329216d446ead17e007

**Review by djrtwo — APPROVED** [review]
okay, I'm fine with it, assuming that clients can no-op functions for a fork without too much issue.

approved

**Pull request merged**
